specs/WebResourceRequested-CustomScheme.md

```diff
@@ -0,0 +1,295 @@
+  // Set if the scheme will be treated as a Secure Context.
+  [propput] HRESULT IsSecure([in] BOOL value);
+
+  // Array of origins that are allowed to issue requests with the custom scheme.
+  // Except origins with this same custom scheme, which are always allowed,
+  // the origin of any request (requests that have the 
+  // [Origin header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin)) to the custom scheme URL
+  // needs to be in this list. No-origin requests are requests that do not have an Origin header,
+  // such as link navigations, embedded images and are always allowed.
+  // Note that cross-origin restrictions still apply. 
+  // If the list is empty, no cross-origin request to this scheme is allowed.
+  // Origins are specified as a string in the format of scheme://host:port.
+  // The origins are string pattern matched with `*` (matches 0 or more characters)
+  // and `?` (matches 0 or 1 character) wildcards just like the URI matching in the
+  // [AddWebResourceRequestedFilter API](https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2.addwebresourcerequestedfilter).
+  // For example, "http://*.example.com:80".
+  // The array of strings and the array here must be deallocated with CoTaskMemFree.
+  HRESULT GetAllowedOrigins([out] UINT32* allowedOriginsCount, [out] LPCWSTR** allowedOrigins);
+  // Set the array of origins that are allowed to use the scheme.
+  HRESULT SetAllowedOrigins([in] UINT32 allowedOriginsCount, [in] LPCWSTR* allowedOrigins);
+}
+
+// This is the ICoreWebView2EnvironmentOptions3 interface
+[uuid(ac52d13f-0d38-475a-9dca-876580d6793e), object, pointer_default(unique)]
+interface ICoreWebView2EnvironmentOptions3 : IUnknown {
+  /// Array of custom scheme registrations. The array must be deallocated with CoTaskMemFre.
+  HRESULT GetCustomSchemeRegistrations(
+      [out] UINT32* count,
+      [out] ICoreWebView2CustomSchemeRegistration*** schemeRegistrations);
+  /// Set the array of custom scheme registrations to be used.
+  HRESULT SetCustomSchemeRegistrations(
+      [in] UINT32 count,
+      [in] const ICoreWebView2CustomSchemeRegistration** schemeRegistrations);
+}
+```
+
+## .NET API
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    // Represents the registration of a custom scheme with the CoreWebView2Environment.
+    // This allows the WebView2 app to be able to handle
+    // WebResourceRequested event for requests with the specified scheme and
+    // be able to navigate the WebView2 to the custom scheme. Once the environment
+    // is created, the registrations are valid and immutable throughout the
+    // lifetime of the associated WebView2s' browser process and any WebView2
+    // environments sharing the browser process must be created with identical
+    // custom scheme registrations, otherwise the environment creation will fail.
+    // Any further attempts to register the same scheme will fail.
```

Comment:
"Fail" means "constructor will Throw an Exception", correct?  Are we considering constructor of an existing scheme user-error?

Reply:
Actually failure occurs when creating the environment with duplicate scheme name. Documented.